File: QA-project/backend/src/db/schema/defects.ts
import {
  pgTable,
  serial,
  integer,
  varchar,
  timestamp,
} from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm";
import { datasets } from "./datasets";
import { defectLifecycle } from "./defect-lifecycle";

export const defects = pgTable("defects", {
  defectId: serial("defect_id").primaryKey(),
  datasetId: integer("dataset_id")
    .notNull()
    .references(() => datasets.datasetId, { onDelete: "cascade" }),
  bugId: varchar("bug_id", { length: 100 }),
  title: varchar("title", { length: 500 }).notNull(),
  module: varchar("module", { length: 255 }),
  severity: varchar("severity", { length: 50 }),
  priority: varchar("priority", { length: 50 }),
  environment: varchar("environment", { length: 100 }),
  status: varchar("status", { length: 50 }),
  createdDate: timestamp("created_date"),
  resolvedDate: timestamp("resolved_date"),
  closedDate: timestamp("closed_date"),
});

export const defectsRelations = relations(defects, ({ one, many }) => ({
  dataset: one(datasets, {
    fields: [defects.datasetId],
    references: [datasets.datasetId],
  }),
  lifecycle: many(defectLifecycle),
}));

File: QA-project/backend/app/models/defects.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.database import Base


class Defect(Base):
    __tablename__ = "defects"

    defect_id = Column(Integer, primary_key=True, autoincrement=True)
    dataset_id = Column(Integer, ForeignKey("datasets.dataset_id", ondelete="CASCADE"), nullable=False)
    bug_id = Column(String(100))
    title = Column(String(500), nullable=False)
    module = Column(String(255))
    severity = Column(String(50))
    priority = Column(String(50))
    environment = Column(String(100))
    status = Column(String(50))
    created_date = Column(DateTime)
    resolved_date = Column(DateTime)
    closed_date = Column(DateTime)

    # Relationships
    dataset = relationship("Dataset", back_populates="defects")
    lifecycle = relationship("DefectLifecycle", back_populates="defect", cascade="all, delete-orphan")
